wlm/src/main/java/ai/djl/serving/wlm/ModelInfo.java

```diff
@@ -588,6 +609,143 @@ private String inferEngine() throws ModelNotFoundException {
+                || prop.containsKey("option.model_id");
+    }
+
+    private String inferLMIEngine() throws ModelException, IOException {
+        // MMS/Torchserve
+        if (Files.isDirectory(modelDir.resolve("MAR-INF"))) {
+            return "Python";
+        }
+        JsonObject modelConfig = getHuggingFaceModelConfig();
+        if (modelConfig == null) {
+            return "Python";
+        }
+
+        String modelType;
+        if (modelConfig.has("_diffusers_version")) {
+            modelType = "stable-diffusion";
+        } else {
+            modelType = modelConfig.get("model_type").getAsString();
+        }
+        long numAttentionHeads = Long.MAX_VALUE;
+        // All of these are valid in the config.json file for the number of attention heads
+        if (modelConfig.has("num_attention_heads")) {
+            numAttentionHeads = modelConfig.get("num_attention_heads").getAsLong();
+        } else if (modelConfig.has("num_heads")) {
+            numAttentionHeads = modelConfig.get("num_heads").getAsLong();
+        } else if (modelConfig.has("n_head")) {
+            numAttentionHeads = modelConfig.get("n_head").getAsLong();
+        }
+        int tensorParallelDegree = CudaUtils.getGpuCount();
+        if (Utils.getEnvOrSystemProperty("TENSOR_PARALLEL_DEGREE") != null) {
+            tensorParallelDegree =
+                    Integer.parseInt(Utils.getEnvOrSystemProperty("TENSOR_PARALLEL_DEGREE"));
+        } else if (prop.get("option.tensor_parallel_degree") != null) {
+            tensorParallelDegree =
+                    Integer.parseInt(prop.get("option.tensor_parallel_degree").toString());
+        }
+
+        if (tensorParallelDegree > 0) {
+            prop.put("option.tensor_parallel_degree", tensorParallelDegree);
+        }
+        String huggingFaceTask = Utils.getEnvOrSystemProperty("HF_TASK");
+        if (huggingFaceTask != null) {
+            prop.put("option.task", huggingFaceTask);
+        }
+
+        if ("stable-diffusion".equals(modelType)) {
+            // TODO: Move this from hardcoded to deduced in PyModel
+            prop.put("option.entryPoint", "djl_python.stable-diffusion");
+            return "DeepSpeed";
+        }
+
+        if (!isTensorParallelSupported(numAttentionHeads, tensorParallelDegree)) {
+            return "Python";
+        }
+        if (isDeepSpeedRecommended(modelType)) {
+            return "DeepSpeed";
+        }
+        if (isFasterTransformerRecommended(modelType)) {
+            return "FasterTransformer";
+        }
+        return "Python";
+    }
+
+    private URI generateHuggingFaceConfigUri(String modelId) throws ModelException, IOException {
+        URI configUri = null;
+        if (modelId != null && modelId.startsWith("s3://")) {
+            // This is definitely suboptimal, but for the majority of cases we need to download this
+            // s3 model eventually, so it is not the worst thing to download it now.
+            downloadS3();
```

Comment:
did you prevent from model redownloading?

Reply:
the `downloadS3` should prevent that already. We check to see if the tmp/<hash> exists before calling s3